source/extensions/filters/http/file_system_buffer/fragment.h

```diff
@@ -0,0 +1,102 @@
+#pragma once
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "envoy/buffer/buffer.h"
+
+#include "source/extensions/common/async_files/async_file_handle.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace FileSystemBuffer {
+
+class FragmentData;
+using UpdateFragmentFunction = std::function<void()>;
+using Extensions::Common::AsyncFiles::AsyncFileHandle;
+using Extensions::Common::AsyncFiles::CancelFunction;
+
+// A Fragment is a piece of the buffer queue used by the filter.
+//
+// Each fragment may be in memory, on disk, or in an unusable intermediate state
+// while controlled by the AsyncFiles library.
+//
+// The fragments are queued in sequential order, and may not be in the same order
+// in the buffer file, as we write "last fragment first" so that the soonest-needed
+// fragments remain in memory as long as possible.
+//
+// Memory fragments are simply `Buffer::Instance`s. Storage fragments contain just
+// enough information to reload a memory fragment from the buffer file.
+class Fragment {
+public:
+  explicit Fragment(Buffer::Instance& buffer);
+  explicit Fragment(Buffer::Instance& buffer, size_t size);
+  ~Fragment();
+
+  // Starts the transition for this fragment from memory to storage.
+  //
+  // The on_done callback is called when the file write completes.
+  //
+  // The callback includes its own callback parameter update_fragment (on success),
+  // so that the operation of updating the fragment can be dispatched to be performed
+  // in the envoy worker thread, rather than being performed on the callback thread.
+  // (This is important to allow for proper handling if the context was deleted while
+  // the operation was in flight, for example.)
+  // i.e. typical usage resembles:
+  // auto result = fragment.toStorage(file_handle, offset,
+  //     [this](absl::StatusOr<UpdateFragmentFunction> status_or_update_fragment) {
+  //   if (status_or_update_fragment.ok()) {
+  //     auto update_fragment = status_or_update_fragment.value();
+  //     dispatcher->dispatch([this, update_fragment]() {
```

Comment:
Where does dispatcher come from in this example? Related to our convo on Slack, `this` can likely also be gone by the time dispatcher runs. Should some kind of weak_ptr callback system be built into this thing so the caller doesn't need to bother? Or can we at least document more how this is supposed to be used?

Reply:
I was trying to keep this helper more context-agnostic, but you're right, it is simpler if the helper gets a dispatch function directly. I still offloaded the safety aspect to the caller, because that way Fragment doesn't need to know anything about filters, but taking a dispatch function parameter makes it much simpler than a callback that returns a callback that must be called back!